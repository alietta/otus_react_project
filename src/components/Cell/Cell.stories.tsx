import React, { FC } from "react";
import { withKnobs, boolean } from "@storybook/addon-knobs";
import { Cell } from "./Cell";

export default {
  title: "Cell",
  decorators: [withKnobs],
};

export const cellNotFilled: FC = () => (
  <Cell isFilled={boolean("isFilled", false)} />
);
export const cellFilled: FC = () => (
  <Cell isFilled={boolean("isFilled", true)} />
);
