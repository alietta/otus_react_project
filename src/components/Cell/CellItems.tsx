import styled from "@emotion/styled";
import { css } from "@emotion/core";

const EmptyCell = css`
  background: white;
`;

const FilledCell = css`
  background: lightgray;
`;

interface Props {
  isFilled: boolean;
}

export const CellWrapper = styled.div`
  display: inline-block;
  border: 1px solid gray;
  box-sizing: border-box;
  vertical-align: middle;
  &:hover {
    border-width: 3px;
  }
  ${({ isFilled }: Props): string => (isFilled ? FilledCell : EmptyCell)};
  ${({ width }: Props): number => `width: ${width}px`};
  ${({ height }: Props): number => `height: ${height}px`};
`;
