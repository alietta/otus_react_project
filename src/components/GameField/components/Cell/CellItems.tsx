import styled from "@emotion/styled";
import { css } from "@emotion/core";

const EmptyCell = css`
  background: white;
`;

const FilledCell = css`
  background: lightgray;
`;

interface Props {
  isFilled: boolean;
}

export const CellWrapper = styled.div`
  display: inline-block;
  width: 30px;
  height: 30px;
  border: 1px solid gray;
  margin: 3px;
  box-sizing: border-box;
  &:hover {
    border-width: 3px;
  }
  ${({ isFilled }: Props): string => (isFilled ? FilledCell : EmptyCell)};
`;
