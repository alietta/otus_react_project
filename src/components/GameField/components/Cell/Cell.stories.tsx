import React from "react";
import { withKnobs, number, array, boolean } from "@storybook/addon-knobs";
import { Cell } from "./Cell";

export default {
  title: "Cell",
  decorators: [withKnobs],
};

export const cellNotFilled = () => (
  <Cell isFilled={boolean("isFilled", false)} />
);
export const cellFilled = () => <Cell isFilled={boolean("isFilled", true)} />;
