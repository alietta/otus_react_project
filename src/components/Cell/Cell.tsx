import React, { FunctionComponent, useMemo } from "react";
import { CellWrapper } from "./CellItems";
import { useTheme } from "sancho";

interface CellProps {
  x: number;
  y: number;
  isFilled: boolean;
  onClick: (x: number, y: number, isFilled?: boolean) => void;
  size: number;
}

const Cell: FunctionComponent<CellProps> = (props: CellProps) => {
  const theme = useTheme();
  const { isFilled, onClick, x, y, size = 5 } = props;
  const fillColor = useMemo(() => {
    const {
      colors: { background, text },
    } = theme;
    return isFilled ? text.default : background.default;
  }, [isFilled]);
  const onCellClick = (): void => {
    onClick(x, y, !isFilled);
  };

  return (
    <CellWrapper
      css={{ borderColor: theme.colors.border.default, background: fillColor }}
      onClick={onCellClick}
      width={size}
      height={size}
    ></CellWrapper>
  );
};

export { Cell };
