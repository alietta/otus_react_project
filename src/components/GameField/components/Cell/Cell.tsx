import React, { FunctionComponent } from "react";
import { CellWrapper } from "./CellItems";

interface CellProps {
  x: number;
  y: number;
  isFilled: boolean;
  onClick: (x: number, y: number, isFilled?: boolean) => void;
}

const Cell: FunctionComponent<CellProps> = (props: CellProps) => {
  const { isFilled, onClick, x, y } = props;
  const onCellClick = (): void => {
    onClick(x, y, !isFilled);
  };

  return <CellWrapper isFilled={isFilled} onClick={onCellClick}></CellWrapper>;
};

export { Cell };
