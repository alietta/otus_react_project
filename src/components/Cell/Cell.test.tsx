/**
 * @jest-environment jsdom
 */
import React from "react";
import { mount, configure } from "enzyme";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
configure({ adapter: new Adapter() });

import { Cell } from "./Cell";

describe("Cell", () => {
  it("renders button for empty cell", () => {
    expect(renderer.create(<Cell onClick={jest.fn()} />).toJSON())
      .toMatchInlineSnapshot(`
      .emotion-0 {
        display: inline-block;
        border: 1px solid gray;
        box-sizing: border-box;
        vertical-align: middle;
        background: white;
        width: 5px;
        height: 5px;
        border-color: hsla(210,10.8%,14.5%,0.12);
        background: white;
      }

      .emotion-0:hover {
        border-width: 3px;
      }

      <div
        className="emotion-0 emotion-1"
        height={5}
        onClick={[Function]}
        width={5}
      />
    `);
  });
  it("renders button for filled with x cell", () => {
    expect(
      renderer.create(<Cell onClick={jest.fn()}>x</Cell>).toJSON()
    ).toMatchSnapshot();
  });
  it("calls onClick callback on click by empty cell", () => {
    const onClick = jest.fn();
    const wrapper = mount(<Cell onClick={onClick} />);
    wrapper.simulate("click");
    expect(onClick).toHaveBeenCalled();
  });
  it("calls onClick callback with passed x, y params", () => {
    const onClick = jest.fn();
    const x = 12;
    const y = 14;
    const wrapper = mount(<Cell onClick={onClick} x={x} y={y} isFilled />);
    wrapper.simulate("click");
    expect(onClick).toHaveBeenCalledWith(x, y, false);
  });
});
