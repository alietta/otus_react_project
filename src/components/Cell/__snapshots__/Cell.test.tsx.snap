// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders button for filled with x cell 1`] = `
.emotion-0 {
  display: inline-block;
  border: 1px solid gray;
  box-sizing: border-box;
  vertical-align: middle;
  background: white;
  width: 5px;
  height: 5px;
  border-color: hsla(210,10.8%,14.5%,0.12);
  background: white;
}

.emotion-0:hover {
  border-width: 3px;
}

<div
  className="emotion-0 emotion-1"
  height={5}
  onClick={[Function]}
  width={5}
/>
`;
